### DISCUSSION TITLE
Pure Python Frontend for Mesa

### DISCUSSION THREAD

**Discussion by ankitk50** [mesa/mesa#1622] [General]
Hi all,

I am looking at a pure python frontend framework to be used for Mesa. So far I'm looking into StreamLit, Dash and Panel. All these are open-source packages. 

For an initial pipe-cleaner/proof of concept I tried to the Boltzmann Money Model example from the documentation. I was able to start the simulation and visualise the results using StreamLit browser based UI . The more details on the approach and the instruction to run the UI can be found here: [github Link](https://github.com/ankitk50/mesa_frontEnd). 

Here is a screenshot: 
<img width="803" alt="Screenshot 2023-03-22 at 15 43 04" src="https://user-images.githubusercontent.com/19813545/227963384-7af32c8a-c44b-4b4f-93ac-c296c5406458.png">

I now try to see the Jupyter compatibility of these packages. Looking forward for your views on this.

**Comment by wang-boyu** [comment]
What a coincidence! I was exploring this very topic yesterday. Mesa models can be run headless without browser, so in theory it should be possible to use any Python frontend for visualization, with of course some extra work linking Mesa models and these frontend frameworks.

What I created yesterday was a very simple visual for the Schelling example. The code is available at [this link](https://github.com/wang-boyu/mesa-examples/blob/feat%2Fstreamlit-frontend/examples/Schelling/st_app.py) and it looks like the following:

<img width="1238" alt="Screenshot 2023-03-27 at 8 45 11 AM" src="https://user-images.githubusercontent.com/10785873/227975917-6a9588f7-c80c-4e99-a6ff-bd020c4c5a9b.png">

If Mesa can provide some View classes for the built-in spaces (grid, continuous, network, etc), then we can re-use lots of widgets/components (slider, button, etc) in Streamlit, and plot whatever diagram we want in matplotlib and show it in Streamlit app.

The downside of it is that there's an extra dependency on Streamlit, and users will need to learn how to use it. But given the ease of use and richness of its documentation, hopefully this won't cause too much trouble.

**Reply by ankitk50** [reply]
The end to end flow would be something like this:

<img width="888" alt="Screenshot 2023-03-27 at 19 32 50" src="https://user-images.githubusercontent.com/19813545/228020961-02ea5f53-6fc7-4878-b337-8882c3f10bc1.png">

The interfacing API or (the "View class" as you mentioned) can be extended to support other frameworks like Dash, Panel as well.

**Reply by wang-boyu** [reply]
Yup. The requirements are pretty much summarized in a [project wiki page](https://github.com/projectmesa/mesa/wiki/Projects-for-Google-Summer-of-Code-2023#additional-ideas) ("Pure python front end" under Additional Ideas). An important task moving forward would be to define this concrete set of interfacing APIs, so that they can be in turn implemented to connect Dash/Panel/Streamlit.

**Reply by wang-boyu** [reply]
Another related (or not) issue that I opened last year: https://github.com/projectmesa/mesa/issues/1441

**Reply by rht** [reply]
I have a hard preference for the frontend to also be runnable inside a Jupyter notebook (so that it is easily runs on Binder, Colab, etc), but I admit that Streamlit UI is definitely slick, and that the framework is wildly popular, that perhaps it would be fine to use it for local self-hosted server instead, and we would have a second frontend written in Panel/ipywidgets for Jupyter support. This is despite https://quansight.com/post/dash-voila-panel-streamlit-our-thoughts-on-the-big-four-dashboarding-tools clearly saying that a Streamlit server doesn't scale to large number of users, and that it doesn't support big data viz either.

**Reply by ankitk50** [reply]
I also think Streamlit is a good place to start working in this direction. We can make a dashboard and try porting our examples to it. Later for Jupyter support we can extend the API to allow calls to panel.

**Reply by rht** [reply]
OK, I think Streamlit for now is fine, so that we can start prototyping the concrete set of APIs early on.

**Reply by rht** [reply]
@wang-boyu do you know why the step takes several seconds in your prototype? Is the model reinitialized from scratch after each step?
Why is the code written this way?
```python
if step_button:
    schelling_model.step()
```
instead of a callback?

**Reply by wang-boyu** [reply]
Interesting - the step function didn't take several seconds to run on my laptop (but definitely slower than current Mesa frontend). The model logic was not changed at all. Only the visualization of the grid was changed.

The code was not written very well as I was simply trying to see how we could visualize grid space in Streamlit, without paying too much attention on how to implement the start/step/stop/reset buttons.

While rendering the grid, the agents were portrayed one by one in a nested loop, which is definitely not a performant way of doing it and perhaps causing the slowness:

```python
portrayal_list = []
for x in range(self.grid.width):
    for y in range(self.grid.height):
        cell_objects = self.grid.get_cell_list_contents([(x, y)])
        for obj in cell_objects:
            portrayal = self.portrayal_method(obj)
            if portrayal:
                portrayal_list.append(
                    {"x": x, "y": y, "Color": portrayal["Color"]}
                )
```

Btw I was using matplotlib to plot the grid space at each step. Wondering if it would be better to implement a custom Streamlit component instead. (Ref: [Create a Component](https://docs.streamlit.io/library/components/create))

**Reply by rht** [reply]
I read about it, and it's actually a limitation of Streamlit that the entire code is always re-executed whenever there is a state update. The model is recreated every time, and there needs an [explicit caching](https://docs.streamlit.io/library/advanced-features/caching) instead. I have searched the internet on whether someone had written a game of life in Streamlit. Couldn't find it. Meanwhile you can readily [find it](https://panel.holoviz.org/gallery/simple/deckgl_game_of_life.html) in Panel gallery of examples. It looks like hash.ai's 3D renderings.

Regarding with using Matplotlib, I think it's not necessarily slower than explicit portrayal UI components, as the latter still needs full redraw during updates anyway. I'm more concerned of the several seconds of delay. Should it come to this, I think custom Streamlit components are much harder to write, as Streamlit is inflexible

**Reply by ankitk50** [reply]
Yes, I also came across this limitation for StreamLit. If you pause the execution of the page in between then there is no way to continue the execution from there, instead the whole page is reloaded. And as you mentioned, the way to make it faster is caching the system states.

**Reply by wang-boyu** [reply]
If the Game of Life demo in Panel was based on PyDeck (as mentioned [in their website](https://panel.holoviz.org/gallery/simple/deckgl_game_of_life.html)), then perhaps it can also be ported to Streamlit using its [`st.pydeck_chart`](https://docs.streamlit.io/library/api-reference/charts/st.pydeck_chart)?

**Reply by ankitk50** [reply]
I can give this a try!

**Reply by ankitk50** [reply]
I tried porting Conway's game of life example in mesa to Streamlit. I just submitted a PR [here](https://github.com/projectmesa/mesa-examples/pull/37 ).

**Comment by ankitk50** [comment]
Hi all,

I also tested [Panel](https://panel.holoviz.org/index.html). It works quite well with Jupyter notebook. 

<img width="814" alt="Screenshot 2023-03-27 at 23 22 58" src="https://user-images.githubusercontent.com/19813545/228074290-8c2dbf5b-feeb-46bb-95b2-0619c9fdf1d8.png">

I use the same example. The source can be found in the [notebook](https://github.com/ankitk50/mesa_frontEnd/blob/main/Panel_Mesa_Example.ipynb)

**Reply by rht** [reply]
I think Panel is relatively more extensible (and that it is 100% FOSS). One way to test this would be to implement the modal popup for shutdown in #1675, in various frameworks. It doesn't make sense to have the shutdown button in a Jupyter notebook, but it does in the stand alone instance.

**Reply by ankitk50** [reply]
I can give it a try out this weekend, thanks for the idea.

**Reply by rht** [reply]
I made a working prototype for the boltzmann_wealth_model example in Panel
https://gist.github.com/rht/e0a63e1cae232a49d8142eb1952c0bd2. To try it, you simply put the gist file in the boltzmann_wealth_model root folder.
![2023-05-10-043947_830x585_scrot](https://github.com/projectmesa/mesa/assets/395821/b7833c85-d85d-4444-993a-efc43120dbcd)

Based on reading https://panel.holoviz.org/about/comparisons.html, in addition to the lack of Jupyter support in Streamlit, a key takeaway:
> Streamlit’s key difference from those other tools is that with streamlit, the entire Python source file is effectively re-run every time a widget changes value, which has the advantage of not allowing confusing out-of-order execution of notebook cells, and also can make it simpler to reason about state in general. However, for this approach to be practical, it requires all lengthy computations to be made cacheable, which is not always straightforward and can introduce its own highly complicated reasoning about state.

However, 2 main issues I have with Panel are:
- It requires Bokeh, which requires huge dependencies installation. I suppose in 2023 people don't care much about storage size, and so I'm willing to relax the condition of Mesa being lean
- The server startup time is several seconds. A bit too slow

An alternative option would be do use plain `ipywidgets` and run it as a standalone application, see https://discourse.jupyter.org/t/ipython-display-in-standalone-application/1048.

**Reply by rht** [reply]
The server startup time can be ruled out because @wang-boyu's Streamlit implementation also takes several seconds to load.

The `ipywidgets` option is essentially Voilà.

**Comment by rht** [comment]
There is also [Voilà](https://voila.readthedocs.io/en/stable/).

**Reply by ankitk50** [reply]
Yes, I checked this one during early analysis and they only allowed publishing Jupyter notebook via a Webserver.

**Reply by rht** [reply]
I created a prototype with the most number of features (among the Streamlit, Panel, etc demo) so far, in Voilà: https://github.com/rht/mesa-examples/tree/voila2.

**Comment by rht** [comment]
Comparison articles:
- https://quansight.com/post/dash-voila-panel-streamlit-our-thoughts-on-the-big-four-dashboarding-tools
- https://medium.datadriveninvestor.com/streamlit-vs-dash-vs-voil%C3%A0-vs-panel-battle-of-the-python-dashboarding-giants-177c40b9ea57#acf9 (has a table that neatly summarizes the 4 frameworks)
- https://panel.holoviz.org/about/comparisons.html (possibly biased because it is written by Panel)

**Reply by rht** [reply]
Install size comparisons from a fresh virtualenv:
- `pip install panel ipywidgets ipywidgets_bokeh` 349.6 MiB: `bokeh` (84.7 MiB), `panel` (75.3 MiB), `numpy` (34 MiB), ...
- `pip install streamlit` 384.4 MiB: `pyarrow` (126.3 MiB), `pandas` (61.6 MiB), `numpy` (34 MiB), ..., `streamlit` (28.3 MiB)
- `pip install ipywidgets voila` 172.0 MiB: `babel` (33.0 MiB), `debugpy` (17.4 MiB), ...

As such, this rules out my concern that Panel has the heaviest dependencies, as Streamlit actually is comparable due to the `pyarrow` dependency.

**Reply by rht** [reply]
`pip install solara` 256.7 MiB: `numpy` (35.3 MiB), `numpy.libs` (34.4 MiB), `solara` (17.5 MiB), ...
The effective size should be smaller, because the largest 2 packages are NumPy's.

**Reply by maartenbreddels** [reply]
Note, Solara is now about ~164MB (well, the soon to be released)

**Reply by rht** [reply]
Why is it far smaller?

**Reply by maartenbreddels** [reply]
Less dependencies by default https://github.com/widgetti/solara/pull/214

pillow, numpy, watchgod, pygments and jupyter-server are removed

**Comment by rht** [comment]
I have this thought in mind but haven't said it: the extra value in Mesa in switching to pure Python frontend is that we may no longer have to deal with custom [JS deps downloads](https://github.com/projectmesa/mesa/blob/e3af2a508080d343eb44f07c5fe5cc9f487a19aa/setup.py#L49-L62). They all can be removed, and we can finally move to declarative pyproject.toml and setup.cfg. (Then finally I can have the bandwidth to grow mesa-examples instead of having to worry about the software engineering stuff.)

**Comment by rht** [comment]
I have seen that https://gradio.app/ is used in the UIs for Hugging Face. @wang-boyu have you looked into it, since it is listed in one of the possible frameworks to use in the GSoC wiki?
See also https://github.com/projectmesa/mesa/discussions/1276.

**Reply by wang-boyu** [reply]
No unfortunately I haven't used it before. Not sure how it's different from Streamlit.

**Reply by ankitk50** [reply]
It's by hugging face and looks like more inclined towards machine learning applications.

**Reply by rht** [reply]
But it can run on Jupyter. Can it do animation?

**Reply by ankitk50** [reply]
I will have to check with Jupyter native, although [here](https://colab.research.google.com/github/gradio-app/gradio/blob/main/demo/live_dashboard/run.ipynb#scrollTo=2.8891853944186117e%2B38) is seems to do animation in colab notebook.

**Reply by rht** [reply]
They have an explicit `update` method, but it's not via existing plotting libraries. Did you do redraw from scratch, or update, in the game of life example?

**Reply by ankitk50** [reply]
I have a base graph object and then I compute the mesh grid and plot it every time on the base layer.

**Comment by tpike3** [comment]
see attached proposal and provide comments and or critiques.  A lot was taken from Ankit's proposal but based on the typeform where it needs to be submitted I was unable to add pictures in my test so no diagram. 


[Mesa NumFOCUS Small Development Grant Application Questions -draft.pdf](https://github.com/projectmesa/mesa/files/11642657/Mesa.NumFOCUS.Small.Development.Grant.Application.Questions.draft.pdf)

**Reply by wang-boyu** [reply]
Thanks Tom for preparing this! The submission deadline is June 16th? Hope we can get it this time. Besides, I think there's a typo in my first name (probably due to grammar checker : ))

**Reply by tpike3** [reply]
Whoops, sorry about that @wang-boyu, yes definitely auto-correct.  I will get that fixed and clean up the language some more as well 

Yes, June 16th is the deadline.

**Reply by ankitk50** [reply]
Hi Tom, very nice proposal. There is a typo in your email address [dot(cm)] (please ignore if that is intentional to avoid web crawlers).

**Reply by tpike3** [reply]
Proposal submitted..... Based on the feedback, I used the enclosed. 
[Submitted-- NumFOCUS Small Development Grant Application Questions.pdf](https://github.com/projectmesa/mesa/files/11715459/Submitted--.NumFOCUS.Small.Development.Grant.Application.Questions.pdf)

**Comment by Corvince** [comment]
Maybe this is also a good time to rethink splitting mesa into different packages? (#1295) With mesa-examples already gone it would be a smaller step to create an own repo for the frontend and leave the remaining bits of mesa as mesa-core. 
This would mitigate the problem of large dependencies coming from the frontend frameworks discussed in this thread. 

It would also allow us to keep the current frontend in a separate repo. I think the current ModularServer has some nice properties that would be a pity to miss. First of all the possibility to create truly custom displays is something I have rarely seen anywhere. Second it is fairly easy to host. You just have to start the server and that's it. 

Separating the repo would also have the advantage that we could build the frontend gradually without worrying about breaking changes for mesa-core. Starting in a pre-v1 environment allows for rapid development, whereas we should be more careful with mesa-core. 

On the topic of choosing a framework, I think it would be very nice to have the ability to trigger an action on clicks inside the visualisations as this would allow you to build more interactive experiences. I don't know which of the proposed frameworks allow this.

**Reply by Corvince** [reply]
And just to be clear: I think a pure python frontend is a very good idea and the advantages of the modular server are very niche. Focusing on a jupyter notebook friendly interface should definitely take precedence.

**Reply by tpike3** [reply]
I absolutely agree, I think the future of ABMs is a rich ecosystem of  components so individuals can rapidly piece together models and grow their emergent phenomena of interest. I have dabbled in this [Mesa Behaviors](https://github.com/projectmesabehaviors) and [Mesa Data](https://github.com/projectmesadata) but the largest challenge is sustaining it.

**Comment by maartenbreddels** [comment]
Hi,

you might also want to consider [Solara](https://github.com/widgetti/solara/) (disclaimer: I'm an author for both Voila and Solara). Voila is focused on executing Jupyter notebooks, and rendering the output, where ipywidgets are the default way to build interactive interfaces. Solara has two parts:
 1. UI part that contains ipywidget-based components with a React-like flavor over it
 2. A server that can execute a notebook or Python script and render a single ipywidget.

The server part can be seen as an alternative to Voila if the focus is to only show a UI (and not notebook output) and will perform better due to having a single process vs 1 kernel/process per request.
Hope that helps in making decisions.

Regards,

Maarten

**Reply by rht** [reply]
It looks awesome! I'm trying it right now.

**Reply by rht** [reply]
The Solara [documentation](https://solara.dev/docs) is organized based on https://diataxis.fr/.

**Reply by maartenbreddels** [reply]
good observation, I try to stay close to that as possible indeed.

**Reply by rht** [reply]
I have made a version in Solara: https://github.com/rht/mesa-examples/blob/voila2/examples/boltzmann_wealth_model/mesa_solara.py. The user-facing code is https://github.com/rht/mesa-examples/blob/voila2/examples/boltzmann_wealth_model/run_solara.py. To run the frontend, simply do `solara run run_solara.py`. The UI is about as slick as Streamlit, yet much more performant. I have implemented step, play, pause -- all the necessary features for it to replace our Tornado-based frontend in the tutorial. We should be able to have a Colab-ready tutorial with this.

We then need to revamp the tutorial, simplify the cookiecutter, etc. Maybe we will have a 3.0 release a few months after 2.0. @ankitk50 if you are up for doing this once we know the NumFOCUS grant result.

**Reply by rht** [reply]
![2023-06-28-105558_854x582_scrot](https://github.com/projectmesa/mesa/assets/395821/32a51aa1-c328-42d2-af43-6d30c6c747ec)

**Reply by ankitk50** [reply]
looks great! I can try out some  more examples.

**Reply by rht** [reply]
It has been merged in the mesa-examples repo: https://github.com/projectmesa/mesa-examples/pull/44, in the folder examples/schelling_experimental and examples/boltzmann_wealth_model_experimental.

Can be readily used via `from mesa_models.experimental import JupyterViz` (to install mesa_models, do `pip install git+https://github.com/projectmesa/mesa-examples`).

On another note, here is a blog post comparing Solara and Streamlit: https://blog.londogard.com/posts/2023-06-30-solara/.

**Reply by ankitk50** [reply]
I would work to integrate Solara with Mesa as part of the next release (3.0). The discussion is started at #1765 .
